### PR TITLE
Scroll blocks into view when mounted

### PR DESCRIPTION
**Pull request by adamsanderson**
This updates Draft's logic around scrolling newly focused blocks into view.  The previous  approach depended on finding the offset values, though those are incorrect when dealing with blocks within a `position: relative` element.  Instead we now get the bounding client rect of the block and the scrolling element and compare them.

This has been tested in both a scrolling element and with window scrolling against paragraphs and list items in Chrome and IE11.

For what it's worth, we have worked around the broken behavior in a variety of places in our code.  Hopefully we can undo some of that logic now.

### CODE REVIEW

**Comment by urbackmk** [comment]
way to tackle some deep draft work!  a couple things:
* looks like you've got a failing test here - i'll wait to review till that is fixed
* i'm new to testing draft-js changes - could you provide some more info on how to test these changes? (both how to get a topic-draft-branch hooked up with textio and also a gif or something of the scrolling behavior that you tested)

File: src/component/contents/DraftEditorBlock.react.js [hunk 1]
```diff
@@ -83,50 +77,49 @@ class DraftEditorBlock extends React.Component {
...
-      scrollDelta = nodeBottom - viewportHeight;
-      if (scrollDelta > 0) {
-        window.scrollTo(
-          scrollPosition.x,
-          scrollPosition.y + scrollDelta + SCROLL_BUFFER,
-        );
-      }
+      // Window itself has a different API for scrolling than elements especially on
+      // IE11.  Luckily we can get the documentElement (<HTML>) and operate on that.
+      scrollElement = window.document.documentElement;
```
Comment (ShannonLCapper) [inline src/component/contents/DraftEditorBlock.react.js]: ⭐️ ⭐️ ⭐️ Browsers don't agree on what the scrollable element is when the whole window scrolls. For some browsers, it's `<html>`, for others, it's `<body>` (see https://makandracards.com/makandra/55801-does-html-or-body-scroll-the-page).

We've addressed this in other parts of the code with a function that returns `document.scrollingElement` (which returns either `<body>` or `<html>` based on the browser implementation) and falls back to `<html>` for IE, which doesn't implement `document.scrollingElement`. https://github.com/textioHQ/frontend/blob/master/packages/webapp/app/es/core/utils/getPageScrollableElement.ts#L37-L40

**Comment by adamsanderson** [comment]
I'll have a editor/webapp branch that pulls this in shortly.  I'm having a *great* time figuring out how to test this.  The Draft tests basically use facebook's helper methods and mock everything out, which doesn't really get the the interesting part of: "Does this work".  

All the same… I'll rework the tests to do something reaosnable.

File: src/component/contents/DraftEditorBlock.react.js [hunk 2]
```diff
@@ -83,50 +77,49 @@ class DraftEditorBlock extends React.Component {
...
     if (!selection.getHasFocus() || endKey !== this.props.block.getKey()) {
       return;
     }
 
-    var blockNode = ReactDOM.findDOMNode(this);
-    var scrollParent = Style.getScrollParent(blockNode);
-    var scrollPosition = getScrollPosition(scrollParent);
-    var scrollDelta;
+    const blockElement = this._element;
+    if (blockElement == null) {
```
Comment (ShannonLCapper) [inline src/component/contents/DraftEditorBlock.react.js]: ⭐️ ⭐️ Why use a double equals here when we could just do a falsy check? `if (!blockElement) { return; }`

**Review by ShannonLCapper — CHANGES_REQUESTED** [review]
I look forward to testing this! Should be a huge improvement

File: src/component/contents/DraftEditorBlock.react.js (hunk 1 above)
Comment (adamsanderson) [reply src/component/contents/DraftEditorBlock.react.js]: Great thanks! Yeah, I was going to use `scrollingElement`, but saw it wasn't available in IE11. When IE11 and Chrome agreed, I assumed I must have covered all possible cases 🤦‍♂.  I'll update that.

File: src/component/contents/DraftEditorBlock.react.js (hunk 2 above)
Comment (adamsanderson) [reply src/component/contents/DraftEditorBlock.react.js]: I just updated that, it came from Draft's implementation.

**Comment by ShannonLCapper** [comment]
This isn't a bug that was introduced in your branch because it happens on dev.textio.tech too, but it looks like we may have some code somewhere that's fighting this code for scroll position. This happens about 75% of the time when I create a new block that's offscreen
Apr-08-2020 10-17-19

It might be a little hard to tell in this video, but here are the steps:
- In a long document, place the cursor at the end of a block towards the bottom of the doc
- Scroll up to the top of the doc so the cursor is offscreen
- Hit enter to create a new block offscreen

Result: It looks like editor scroll position immediately updates to put the new block in view, but then the editor scroll position bounces upwards quickly so that the cursor is offscreen. The amount that it bounces upwards seems to vary. Sometimes it's only one line height's worth, sometimes it's more

**Comment by adamsanderson** [comment]
Shannon: is this against my draft build? 
I have not seen this, but yes there are several things that fight for scroll positioning.  I know of at least two :| Though I haven't seen them.

**Comment by ShannonLCapper** [comment]
More unexpected behavior related to scrolling blocks into view, both not unique to this branch:
1. Hitting enter at the end of an offscreen block in Chrome's mobile simulator doesn't scroll the newly-created block into view. The editor doesn't scroll at all.
2. On iOS (in Simulator), hitting enter at the end of an offscreen block scrolls the window way too high, to the point where sometimes the cursor actually gets put offscreen in the other direction:
Apr-08-2020 10-29-31

**Comment by ShannonLCapper** [comment]
> Shannon: is this against my draft build?

Yes it's a local build that's symlinked to your draft build. But it happens on dev.textio.tech too

File: src/component/utils/scrollElementIntoView.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  }
+
+  const elementRect = element.getBoundingClientRect();
+  const scrollParent = Style.getScrollParent(element);
+  let scrollElement;
+  let scrollRect;
+
+  if (scrollParent === window) {
+    // Window has a different API for scrolling than elements especially on
+    // IE11.  Luckily we can get the documentElement (<HTML>) and operate on that.
```
Comment (ShannonLCapper) [inline src/component/utils/scrollElementIntoView.js]: ⭐️ This comment is slightly out-of-date now

File: src/component/utils/scrollElementIntoView.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  if (elementRect.top < scrollRect.top) {
+    // If the top of the block is above the scroll element, scroll up.
+    scrollElement.scrollTop -= (scrollRect.top - elementRect.top + SCROLL_BUFFER);
+  } else if (elementRect.bottom > scrollRect.bottom) {
+    // If the bottom of the block is below the scroll element, scroll down.
+    scrollElement.scrollTop += (elementRect.bottom - scrollRect.bottom + SCROLL_BUFFER);
+  }
+}
+
+module.exports = scrollElementIntoView;
```
Comment (ShannonLCapper) [inline src/component/utils/scrollElementIntoView.js]: ⭐️ Newline?

**Review by ShannonLCapper — APPROVED** [review]

**Comment by adamsanderson** [comment]
Regarding the bugs that aren't unique to this branch, I have a handful of changes all stacked up behind this.  How would you feel about opening tickets for the ones on mobile?

I think the technique we have now generalizes to getting the viewport as well, or could be used ONLY on browsers that don't already support `scrollIntoViewIfNeeded` (IE maybe firefox) which just says: "Browser… heck, you figure it out".

File: src/component/utils/scrollElementIntoView.js (hunk 4 above)
Comment (adamsanderson) [reply src/component/utils/scrollElementIntoView.js]: √ Added

**Comment by ShannonLCapper** [comment]
Yep I can make tickets for the bugs! Nothing blocking this PR since they're all on dev.textio.tech too

**Pull request merged**